### PR TITLE
Backend endpoints ingredients

### PR DESCRIPTION
**Pull request by Brianna-Flo**
NOTE: Built off of backend-endpoints branch

## Description

<what this pull request is doing>

- creating the backend endpoints for ingredients and integrating with the frontend
- Allow a user to create an ingredient
- Allow a user to edit an ingredient
- Allow a user to delete an ingredient
- Prevent repeat ingredients by mapping multiple users to a single ingredient and keeping the user specific data (quantity and expiration date) separate

<what will be done in later PRs and not included here>

- possible error handling for backend, currently send res.status(500), res.status(404), etc

## Milestones
<the milestones or stories/features that this works towards>

- Integrating the backend and frontend
- allowing a user to input the ingredients and quantity they have on hand

## Resources
<links to tutorials, code snippets, inspirations>
<if you took or translated specific code from the internet, please call it out here!>
https://www.prisma.io/docs/orm/prisma-client/special-fields-and-types/working-with-composite-ids-and-constraints
https://www.prisma.io/docs/orm/prisma-client/queries/relation-queries

## Test Plan
<insert images or gifs of feature>
<otherwise, say how code was tested if not UI facing>
<any edge cases you might have specifically tested>
<div>
    <a href="https://www.loom.com/share/4c40589e4e4b4758be9811b05bfb87c4">
      <p>Add, edit, delete ingredient demo - Watch Video</p>
    </a>
    <a href="https://www.loom.com/share/4c40589e4e4b4758be9811b05bfb87c4">
      
    </a>
  </div>

### CODE REVIEW

File: backend/routes/ingredientsRoutes.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+const router = express.Router()
+
+const { PrismaClient } = require("@prisma/client");
+const prisma = new PrismaClient();
+
+// retrieve all ingredients in the database
+router.get("/", async (req: Request, res: Response) => {
+    try {
+        const ingredients = await prisma.Ingredient.findMany()
+        res.json(ingredients)
```
Comment (paulshaoyuqiao) [inline backend/routes/ingredientsRoutes.ts]: not needed for now, but I highly encourage you to consider adding pagination support for this endpoint (something to do in the next PRs).

Consider a database with hundreds and thousands of ingredients -- if we retrieve them all at once, latency would be really high and the payload would be massive if we fetch them in one go, and would likely cause the website to freeze/crash. 

We commonly use a technique called pagination to divide the response into chunks and send them back iteratively to the app. One easy way to add pagination is to:
1. index your data with some auto-incrementing field (starting at 0)
2. require your endpoint to include 2 new parameters that indicate -- which index to start reading the data from (startIndex), and how many rows to read (pageSize)
3. fetch the data accordingly based on the start index and the page size, and return back to client (i.e. your frontend) the data + the next available index to start fetching from again (or provide a negative number if backend has run out of things to fetch) 
4. Typically, the request to fetch a "new" page of data gets triggered if the user scrolls down the page --> this signals an intent to load more. To keep things even simpler, even just a change from fetching everything in 1 request --> an iterative fetch containing multiple requests would help improve your app's performance and reliability.

File: frontend/src/utils/firebase.ts [hunk 2]
```diff
@@ -30,4 +30,16 @@ const errorCodes: { [errorMessage: string]: string } = {
 const app = initializeApp(firebaseConfig);
 const auth = getAuth(app);
 
-export { auth, errorCodes };
+const getCurrentUserToken = async () => {
+  const user = auth.currentUser
+  
+  if (!user) {
+    // TODO User not logged in error state instead of throw error
+    throw new Error ("user not logged in")
```
Comment (paulshaoyuqiao) [inline frontend/src/utils/firebase.ts]: let's not use `throw` directly.

**Review by paulshaoyuqiao — CHANGES_REQUESTED** [review]
